File: 20_valid_parentheses/isValid.py
class Solution(object):
    def isValid(self, s):
        if len(s) % 2 == 1: return False
        stack = []
        para_dic = {')':'(', ']':'[', '}':'{'}
        for idx in xrange(len(s)):
            if s[idx] in {'(', '[', '{'}:
                stack.append(s[idx])
            else:
                if not len(stack) or para_dic[s[idx]] != stack[-1]: return False
                else: stack.pop()


        return len(stack) == 0

sol = Solution()
s = "{(})[]"
print sol.isValid(s)

File: 20_valid_parentheses/isValid.cpp
class Solution {
public:
    bool isValid(string s) {
        if(s.empty()) return true;
        map<char, char> pmap = {{'{', '}'}, {'[', ']'}, {'(', ')'}};
        //pmap[')'] = '('; pmap[']'] = '['; pmap['}'] = '{';
        stack<char> st;
        for(int i = 0; i < s.size(); i++) {
            if(s[i] == '(' || s[i] == '[' || s[i] == '{') 
                st.push(s[i]);
            else {
                if(st.empty() || st.top() != pmap[s[i]]) return false;
                else st.pop();
            }
        }
        return st.empty();
    }
};